### PR TITLE
feat: gRPC management api stubs

### PR DESCRIPTION
**Pull request by tustvold**
Builds on #846 

This additionally rejigs the proto files such that:

- the module topology within Rust matches the Protobuf package topology - prost assumes this, but hasn't been an issue as we currently lack any cross-package references
- moves the proto definitions into a directory hierarchy under `protos` - this allows setting the root as a search directory when using tools like grpcurl

### CODE REVIEW

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+      ColumnFormat regex = 4;
+      ColumnFormat strfTime = 5;
+    }
+  }
+
+  repeated Part parts = 1;
+}
+
+message Matcher {
+  // Restrict selection to a specific table or table's specified by a regex
```
Comment (carols10cents) [inline generated_types/protos/influxdata/iox/management/v1/database_rules.proto:35]: Small grammar fix:

```suggestion
  // Restrict selection to a specific table or tables specified by a regex
```

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+  // into a partition).
+  repeated string readOnlyPartitions = 4;
+}
+
+message WalBufferConfig {
+  enum Rollover {
+    ROLLOVER_UNKNOWN = 0;
+
+    // Drop the old segment even though it hasn't been persisted. This part of
+    // the WAl will be lost on this server.
```
Comment (carols10cents) [inline generated_types/protos/influxdata/iox/management/v1/database_rules.proto:126]: Small typo:

```suggestion
    // the WAL will be lost on this server.
```

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    ROLLOVER_RETURN_ERROR = 3;
+  }
+
+  // The size the WAL buffer should be limited to. Once the buffer gets to
+  // this size it will drop old segments to remain below this size, but
+  // still try to hold as much in memory as possible while remaining
+  // below this threshold
+  uint64 bufferSize = 1;
+
+  // WAL segments become read only after crossing over this size. Which means
```
Comment (carols10cents) [inline generated_types/protos/influxdata/iox/management/v1/database_rules.proto:145]: I think there's supposed to be a hyphen in here:

```suggestion
  // WAL segments become read-only after crossing over this size. Which means
```

Otherwise it sounds like you can only read WAL segments after they exceed `segmentSize`?

**Review by carols10cents — APPROVED** [review]
Just a few little documentation fixes; the new organization looks fine to me!

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+import "influxdata/iox/management/v1/base_types.proto";
+
+// `PartitionTemplate` is used to compute the partition key of each row that
+// gets written. It can consist of the table name, a column name and its value,
+// a formatted time, or a string column and regex captures of its value. For
+// columns that do not appear in the input row, a blank value is output.
+//
+// The key is constructed in order of the template parts; thus ordering changes
+// what partition key is generated.
+message PartitionTemplate {
```
Comment (alamb) [inline generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: @tustvold  is it fair to say that these are basically "1 to 1 translations" of the Rust `struct`s from the `data_types` crates, such as https://github.com/influxdata/influxdb_iox/blob/main/data_types/src/database_rules.rs#L21?

If so that is cool, I just wanted to make sure.

It would help me review this PR if you could point out any significant deviation (or confirm this is a mechanical translation)

**Review by alamb — APPROVED** [review]
The definitions / rejiggering makes sense to me.

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 4 above)
Comment (tustvold) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: The only conscious "deviation" between the Rust types and the Protobuf types is that the Protobuf implementation lifts the query, subscription and replication configuration out into separate messages, instead of having them directly on the DatabaseRules "structure". This was in part to make the API more consistent, but also to make the type less monolithic and potentially easier to parse, evolve, etc... I should have called this out in the PR description though, so thanks for calling it out :smile: 

The only other change is that the type `PartitionSortRules` is renamed `PartitionDropOrder` to match the field name.

There are, however, some other subtleties to mapping between the types that are worth calling out:

`oneof` is the closest representation of Rust's structured enumerations, but is always optional:

* Where the semantics of not specifying a value made sense, at least to me, I used this "null value" to represent one of the enumerations values, e.g. the protobuf version of MatchTables doesn't have an `All` variant, as this is instead represented by not specifying a value for `tableMatcher` on the owning `Matcher` message.
* Where I felt the semantics of the optional was not clear, the conversion logic in #848 will simply use a default value if one exists (e.g. `MutableBufferConfig.sort`), or require the field to be specified (e.g. `PartionTemplate.Part.part`)

A similar predicament faces the non-structured enumerations, where the 0 variant is always a default value. To workaround this, in all cases the 0 variant is assigned to `<PREFIX>_UNKNOWN`. The logic in #848 then uses a default if there is one defined for the Rust type, otherwise returning an error if the field is not specified.

In various places the Rust types use `Option<String>`, however, prost (and I think Protobuf), doesn't distinguish between an empty string and no string. A strict translation of this would therefore use `google.protobuf.StringValue` instead of `String` but the cases where this showed up I didn't think `Some("")` actually made sense, and so I think the Rust type is actually just overly expressive.

`WalBufferConfig.closeSegmentAfter` is a Duration but `MutableBufferConfig.persistAfterColdSeconds` is an integer, this is an inconsistency in the Rust types that I preserved, but is possibly something we will want to fix as part of the API stabilisation.

**Review by mkmik — COMMENTED** [review]
protobuf style guide says field names should be snake case (e.g. `persist_after_cold_seconds`)

**Comment by tustvold** [comment]
@mkmik right you are, I must have only read the first sentence and missed that it went on to say something different for field names than message names. Will fix (prost does its own thing anyway so it isn't a big deal to change)

File: generated_types/protos/influxdata/iox/management/v1/service.proto [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import "influxdata/iox/management/v1/database_rules.proto";
+
+service Management {
+  rpc Ping(google.protobuf.Empty) returns (google.protobuf.Empty);
+
+  rpc GetWriterId(google.protobuf.Empty) returns (GetWriterIdResponse);
+
+  rpc UpdateWriterId(UpdateWriterIdRequest) returns (google.protobuf.Empty);
+
+  rpc ListDatabases(google.protobuf.Empty) returns (ListDatabasesResponse);
```
Comment (mkmik) [inline generated_types/protos/influxdata/iox/management/v1/service.proto:14]: `google.protobuf.Empty` is wire-protocol compatible with an empty message for any other type, so it's true
that even if we ever decide to add a parameter to `ListDatabases` we can create a `ListDatabaseRequest` message later and retain compatibility with existing clients.

However, we'd break clients who use generated stubs.

A commonly held rule of thumb in these cases is:

"If in doubt, create an empty request message type. If the semantics of the request or response are _by definition_ such that a request or response is only a sentinel/marker (such as a response to a delete operation) then feel free to use Empty."

In particular, the response to a create operation could easily include information about the object being created, so it's not a good candidate for Empty (even if you don't need any field right now).

I found some references online for that rule of thumb:

https://www.youtube.com/watch?v=F2WYEFLTKEw (slides https://static.sched.com/hosted_files/kccncna17/ad/2017%20CloudNativeCon%20-%20Mod%20gRPC%20Services.pdf)

File: generated_types/protos/influxdata/iox/management/v1/base_types.proto [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+enum ColumnType {
+  COLUMN_TYPE_UNKNOWN = 0;
+  COLUMN_TYPE_I64 = 1;
+  COLUMN_TYPE_U64 = 2;
+  COLUMN_TYPE_F64 = 3;
+  COLUMN_TYPE_STRING = 4;
+  COLUMN_TYPE_BOOL = 5;
+}
+
+message HostGroup {
```
[the remark this answers is not in the document]
Comment (tustvold) [reply generated_types/protos/influxdata/iox/management/v1/base_types.proto:25]: Correct

File: generated_types/protos/influxdata/iox/management/v1/service.proto [hunk 7]
```diff
@@ -0,0 +1,43 @@
+syntax = "proto3";
+package influxdata.iox.management.v1;
+
+import "google/protobuf/empty.proto";
+import "influxdata/iox/management/v1/database_rules.proto";
+
+service Management {
+  rpc Ping(google.protobuf.Empty) returns (google.protobuf.Empty);
```
[the remark this answers is not in the document]
Comment (tustvold) [reply generated_types/protos/influxdata/iox/management/v1/service.proto:8]: I just wanted a quick and dirty endpoint to test the gRPC plumbing whilst I was getting it all working - will remove

**Comment by tustvold** [comment]
Making draft whilst I placate [buf](https://buf.build/)

**Comment by domodwyer** [comment]
> Making draft whilst I placate [buf](https://buf.build/)

I was just about to suggest Buf.... 👍

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+      string columnName = 1;
+      ColumnType columnType = 2;
+      Aggregate columnValue = 3;
+    }
+
+    // Sort partitions by this order. Last will be dropped first.
+    Order order = 1;
+
+    // Configure sort key
+    oneof sort {
```
Comment (domodwyer) [inline generated_types/protos/influxdata/iox/management/v1/database_rules.proto:185]: I've never seen `Empty` used as the field type in a oneof - nice idea. 

I'm interested to see how this maps to rust - I suspect it might be rather nice!

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+  oneof tableMatcher {
+    string table = 1;
+    string regex = 2;
+  }
+  // A query predicate to filter rows
+  string predicate = 3;
+}
+
+message ReplicationConfig {
+  message Replication {
```
Comment (domodwyer) [inline generated_types/protos/influxdata/iox/management/v1/database_rules.proto:45]: I like the idea but this appears unused?

There's a couple `string` types, ideally we'd type them (as I guess you're planning on here!) or document the expected format - examples include `SubscriptionConfig.Subscription.hostGroupId`, `QueryConfig.primary/secondary/etc` and `Matcher.predicate`

**Review by domodwyer — COMMENTED** [review]
Looks good - I think linting with [Buf](https://docs.buf.build/) will really help here, there's a few patterns that if we get right now save us headaches in the future 👍 

Conductor has a buf config file if you're looking for something to get going. It enforces JSON wire compat, and I'm unsure if we would expect to maintain JSON wire compatibility for IOx (it's quite restrictive, but we had a big chat about JSON gateways) once we release a stable version - definitely worth discussing at standup one day this week.

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 4 above)
Comment (domodwyer) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: Sounds good - I like the `<PREFIX>_UNKNOWN` pattern.

I'm interested to see how all this is applied against a real structure (as in, which fields default to what values) but happy to wait until the PR that contains it is ready :)

The main pain-points we might want to keep an eye on (that I can think of) for default values are:

* The obvious; changing default values is a breaking API change (clients expect X when they don't set a field now get Y)
* Adding new fields to enums becomes a breaking change - older servers don't understand what variant is being sent to them, so they may map to a "default" case which has different behaviour than the client asked for (and the reverse - servers sending responses to old clients)
* An old client that sends a request without a newer field does not fully understand the implications of what it has requested - this was the source of a nasty security bug in a previous job when a default value published something the client intended to be secret.

These are definitely not insurmountable and I'm sure for many fields they're simply not a concern - just worth keeping in mind. I suppose we should also document the default values somewhere and try to keep them in-sync.

Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: yeah, protobuf schema evolution requires care; in particular servers must always assume there will be clients talking with a previous generation schema. The only way to prevent that is by explicitly setting a cutoff (e.g. with a version field or with a versioned service/method name).

I'm a bit confused about the situation with unknown enum variants in rust. Languages with open enum types such as  Go and C++, retain the underlying numeric value, while languages with closed enum types such as Java generate a `UNRECOGNIZED` enum variants. I need to peek at how the rust library generates stubs...

Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: From prost [doc](https://github.com/danburkert/prost#enumerations):
> Enumerations
All .proto enumeration types convert to the Rust i32 type. Additionally, each enumeration type gets a corresponding Rust enum type, with helper methods to convert i32 values to the enum type. The enum type isn't used directly as a field, because the Protobuf spec mandates that enumerations values are 'open', and decoding unrecognized enumeration values must be possible.

Comment (domodwyer) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: I mention it because it is implementation specific and we expect there to be many languages using the API - I suspect rust will map it to some `INVALID` value due to the closed nature of rust enums but we should definitely check - I'm interested in what you find!

Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: https://gist.github.com/mkmik/bde76e12ed891e290dc492a2d9eb7071

TL;DR:

* fields of protobuf enum types become `i32` fields in the generated struct.
* each enum type has an associated rust enum
* the associated rust enum type implements a `from_i32` method and also `From for i32` so you can `.into()` it into where a i32 is expected (e.g. in the aforementioned fields ^^^).

Comment (tustvold) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: So Rust preserves enums as i32 and the `prost::Message` macro adds a helper method with the same name that can be used for getting a converted representation.

The relevant generated code is

```
pub fn column_type(&self) -> super::super::ColumnType {
    super::super::ColumnType::from_i32(self.column_type)
        .unwrap_or(super::super::ColumnType::default())
}

pub fn from_i32(value: i32) -> ::core::option::Option<ColumnType> {
    match value {
        0 => ::core::option::Option::Some(ColumnType::Unspecified),
        1 => ::core::option::Option::Some(ColumnType::I64),
        2 => ::core::option::Option::Some(ColumnType::U64),
        3 => ::core::option::Option::Some(ColumnType::F64),
        4 => ::core::option::Option::Some(ColumnType::String),
        5 => ::core::option::Option::Some(ColumnType::Bool),
        _ => ::core::option::Option::None,
    }
}
```
So if it receives a type it doesn't recognize it will map it to `ColumnType::Unspecified` i.e. the default variant. So a variant it doesn't recognize will get treated as unspecified, which in all cases for the enums results in a server error.

The thought has occurred to me that using the default value for Matcher is problematic, as it will default a "variant" it doesn't recognise, which could cause some fun behaviour - I will fix this by making it explicit

Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: uh, I didn't notice the generated accessor method; I'm tempted to state that this behaviour should be considered harmful, borderline irresponsible; if users wants such a behaviour they should opt-in.

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 8 above)
Comment (tustvold) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:185]:
```
pub enum Sort {
    /// The last time the partition received a write.
    #[prost(message, tag = "2")]
    LastWriteTime(::google_types::protobuf::Empty),
    /// When the partition was opened in the mutable buffer.
    #[prost(message, tag = "3")]
    CreatedAtTime(::google_types::protobuf::Empty),
    /// A column name, its expected type, and whether to use the min or
    /// max value. The ColumnType is necessary because the
    /// column can appear in any number of tables and be of
    /// a different type. This specifies that when sorting
    /// partitions, only columns with the given name and type
    /// should be used for the purposes of determining the partition
    /// order. If a partition doesn't have the given column
    /// in any way, the partition will appear at the
    /// beginning of the list with a null value where all
    /// partitions having null for that value will then be
    /// sorted by created_at_time desc. So if none of the partitions in
    /// the mutable buffer had this column with this type,
    /// then the partition that was created first would
    /// appear last in the list and thus be the first up to
    /// be dropped.
    #[prost(message, tag = "4")]
    Column(ColumnSort),
}
```

Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:185]: /me too lazy to check if Empty implements Default, but that could save a `use`

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 9 above)
Comment (tustvold) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto]: Good point :grin: - I think the strings are placeholders as that part of the codebase has not yet been implemented. I'll revisit once it is clearer what the types should be. Until then I've removed the unused Replication message

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 4 above)
Comment (alamb) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: I really like the idea of proactively adding a "client version" number in the messages / protocol and have the IOx server explicitly check / reject any messages from a previous versions of the client.

This would fall under "defensive programming": ensure we never tried to process "old" messages until such time as we are sure it is important and have proper testing for it.  Without such a mechanism, we'll end up code that appears to properly process arbitrary versions of client messages but we won't have a great story for ensuring the server does something reasonable for those cases. 

This would let us  punt all sorts of potential compatibility questions and code (with a big 🔨  for sure) until such time as it was actually important to spend time on it. 

Just a thought

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 8 above)
Comment (domodwyer) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:182]: I guess I was wrong....

**Comment by domodwyer** [comment]
It looks like this has been lint-nudged along! 

Perhaps we should check in the buf config so it stays up to date as the rest of us expand the API (and create a CI ticket to enforce it).

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 4 above)
Comment (tustvold) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: I don't think we need to take such a brute force approach, at least until we start cutting releases, we can and I intend to just break the API on a whim. It is useful to discuss these things to identify any issues we might encounter in the future, but I at least have no intention of spending a significant amount of time making sure we are maximally forward/backward compatible - I'm not sure it would even make sense to whilst some of the API types are just placeholders (e.g. Matcher.predicate).

Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: more thoughts:

1. we can always add a versioning field later if needed
2. which of the many changes we'd do to our API layer would require us to bump the version?
3. if we need to add versioning, I'd rather put it in the service name / url rather than in the message payload itself; much better ergonomics both for programmatic clients and interactive `grpcurl`
4. if versioning changes too often, people will just bump the version number and things may even work, until they don't. 

I once heard a nice rule of thumb: if you're breaking backward compatibility in an API, that's effectively another API, and if it's another API then it should have a different name (in a way, putting the version number in the API url/name, is a mechanistic flavour of that rule of thumb). This applies to method names but also to message field names: if your semantics for the "name" field changes, you keep the old `name` field around with the old semantics and create a new `better_name` field with the new semantics.

**Comment by tustvold** [comment]
@domodwyer Yeah I started working on that, but got mad at Github Actions and so am taking a little break - I will revisit it tomorrow :sweat_smile:

File: generated_types/protos/influxdata/iox/management/v1/database_rules.proto (hunk 4 above)
Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: >I once heard a nice rule of thumb [..]

Here is some relevant literature:

* https://research.swtch.com/vgo-import#functional_names_immutable_meanings)
* https://www.youtube.com/watch?v=oyLBGkS5ICk

Comment (alamb) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: I think adding a versioning field if we need it later is fine.

What I have seen is that the first time someone wants to make a semantically backwards incompatible change (e.g. add a new required field) they will add "in situ" upgrade logic to try and adapt old requests to the new format. 

Over time this results in data format migration scattered around the code and it is hard to reason about "what is the schema of valid objects that is accepted"

> if we need to add versioning, I'd rather put it in the service name / url rather than in the message payload itself; much better ergonomics both for programmatic clients and interactive grpcurl

I am willing to bet a bottle of reasonable wine (by US 🍷  standards) that such a scheme will result in the version number being changed ~ never  in practice.

Comment (mkmik) [reply generated_types/protos/influxdata/iox/management/v1/database_rules.proto:15]: > I am willing to bet a bottle of reasonable wine (by US 🍷 standards) that such a scheme will result in the version number being changed ~ never in practice.

that entirely depends on how hard is to change; if it's just about changing one line of code and not 100 files perhaps we could do it.

> Over time this results in data format migration scattered around the code and it is hard to reason about "what is the schema of valid objects that is accepted"

yes; it happens; it's all about balancing where the pain is; one one hand you have "just a coding problem", on the other you have the problem of coordinating potentially hundreds of servers running with different versions, rolling upgrades, rollbacks, canaries, ...

**Review by domodwyer — APPROVED** [review]
Proto looks good - interested to see how this plays out w.r.t the thread above!

**Comment by alamb** [comment]
🎉
